### PR TITLE
Remove trailing slashes in toolchain name

### LINKED ISSUE #2873: Issues with `rustup toolchain uninstall <toolchain>/` with trailing slash
Given the existence of a custom toolchain, `toolchain-a`, `rustup toolchain uninstall toolchain-a/` (note trailing `/`) will not delete the symbolic link at `~/.rustup/toolchains/toolchain-a`, which is expected, but will produce an error and delete the contents of the directory linked to `toolchain-a`, which is unexpected.

```
$ ls ~/rust/toolchains/foo
bin  lib
$ rustup toolchain link toolchain-a ~/rust/toolchains/foo
$ rustup toolchain uninstall toolchain-a/
info: uninstalling toolchain 'toolchain-a/'
error: could not remove 'install' directory: '/home/tgnottingham/.rustup/toolchains/toolchain-a/': Not a directory (os error 20)
$ ls ~/rust/toolchains/foo
$
```

I might expect it to say something like this instead:

```
$ rustup toolchain uninstall toolchain-a/
info: no toolchain installed for 'toolchain-a/'
        Did you mean 'toolchain-a'?
```

For non-custom toolchains, the problem is different. As far as I can tell, it doesn't delete anything, but it does produce an error:

```
$ rustup toolchain uninstall nightly-2021-10-10-x86_64-unknown-linux-gnu/
info: uninstalling toolchain 'nightly-2021-10-10-x86_64-unknown-linux-gnu/'
error: could not remove 'update hash' file: '/home/tgnottingham/.rustup/update-hashes/nightly-2021-10-10-x86_64-unknown-linux-gnu/': could not remove 'update hash' file: '/home/tgnottingham/.rustup/update-hashes/nightly-2021-10-10-x86_64-unknown-linux-gnu/': Not a directory (os error 20)
```

However, using an abbreviated name of the toolchain plus a `/`, e.g. `nightly-2021-10-10/`, fails more or less the way I'd expect:

```
$ rustup toolchain uninstall nightly-2021-10-10/
info: no toolchain installed for 'nightly-2021-10-10/'
```

(Though, here also, a `Did you mean 'nightly-2021-10-10'?`, would be a nice bonus feature.)

---

Output of `rustup --version`:
```
rustup 1.24.3 (ce5817a94 2021-05-31)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.55.0 (c8dfcfe04 2021-09-06)`
```
Output of `rustup show`:
```
Default host: x86_64-unknown-linux-gnu
rustup home:  /home/tgnottingham/.rustup

installed toolchains
--------------------

stable-x86_64-unknown-linux-gnu (default)
beta-x86_64-unknown-linux-gnu
nightly-2021-10-10-x86_64-unknown-linux-gnu
nightly-x86_64-unknown-linux-gnu
toolchain-a

installed targets for active toolchain
--------------------------------------

wasm32-unknown-unknown
x86_64-unknown-linux-gnu

active toolchain
----------------

stable-x86_64-unknown-linux-gnu (default)
rustc 1.55.0 (c8dfcfe04 2021-09-06)
```

### PR DESCRIPTION
**Pull request by sathwikmatsa**
This PR fixes #2873

### CODE REVIEW

File: tests/mock/clitools.rs [hunk 1]
```diff
@@ -637,8 +637,7 @@ where
     let mut vars: HashMap<String, String> = HashMap::default();
     self::env(config, &mut vars);
     vars.extend(env.iter().map(|(k, v)| (k.to_string(), v.to_string())));
-    let mut arg_strings: Vec<Box<str>> = Vec::new();
-    arg_strings.push(name.to_owned().into_boxed_str());
+    let mut arg_strings: Vec<Box<str>> = vec![name.to_owned().into_boxed_str()];
```
Comment (kinnison) [inline tests/mock/clitools.rs]: This change is unrelated to this piece of work, please drop the "ok clippy" commit, it's not necessary for this.

**Review by kinnison — CHANGES_REQUESTED** [review]
There's a commit which needs removing; also also we need a test to demonstrate the issue and then demonstrate it is resolved by this patch.  If you need some inspiration, check out cli-v2.rs which should have plenty of example tests you can crib from.

Thank you for your effort.

**Review by 0xPoe — APPROVED** [review]
LGTM. But could you please squash your commits?

**Review by kinnison — APPROVED** [review]
This looks good, sorry for the delay

**Pull request merged**